changefeedccl: add updated timestamps to enriched envelope source

KeithCh opened the pull request:
changefeedccl: add updated timestamps to enriched 
envelope source

When creating an enriched changefeed with the updated option, we'll 
populate ts_ns and ts_hlc in payload.source

Resolves #139661
Epic: CRDB-8665

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143592)
<!-- Reviewable:end -->

asg0451 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -4460,7 +4462,128 @@ func TestChangefeedEnrichedSourceWithData(t *testing.T) {
 			}
 		})
+
+		testutils.RunTrueAndFalse(t, "ts_{ns,hlc}", func(t *testing.T, withUpdated bool) {
```
can you merge this with the above test block, similar to withMVCCTS?

KeithCh replied:
Made the subtests nested

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -108,6 +108,12 @@ func newEnrichedSourceProvider(
 		nonFixedDataIdx[fieldNameMVCCTimestamp] = len(nonFixedJSONFields) - 1
 	}
+	if opts.UpdatedTimestamps {
+		nonFixedJSONFields = append(nonFixedJSONFields, fieldNameUpdatedTSNS)
```
nit: can you make a little helper function for this construction?

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -137,6 +143,14 @@ func (p *enrichedSourceProvider) GetJSON(updated cdcevent.Row) (json.JSON, error
 	}
+	if p.opts.updated {
+		ts := updated.MvccTimestamp
+		if updated.SchemaTS.WallTime > ts.WallTime {
+			ts = updated.SchemaTS
+		}
```
might be worth passing in the encoder's event context for this rather than redoing this logic. see https://github.com/asg0451/cockroach/blob/721f46ea60b7ea3811ce62ef61ff8b0ced1b35fd/pkg/ccl/changefeedccl/encoder_json.go#L452

in that case, probably want to do it for mvccts as well

ditto for avro

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -290,6 +314,28 @@ var allFieldInfo = map[string]fieldInfo{
+	fieldNameUpdatedTSNS: {
+		avroSchemaField: avro.SchemaField{
+			Name:       fieldNameUpdatedTSNS,
+			SchemaType: []avro.SchemaType{avro.SchemaTypeNull, avro.SchemaTypeString},
```
i believe ts_ns has to be an int64

KeithCh committed:
changefeedccl: add updated timestamps to enriched
envelope source

When creating an enriched changefeed with the updated option, we'll
populate ts_ns and ts_hlc in payload.source

Resolves #139661
Epic: CRDB-8665

Release note: none

aerfrei commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -4350,114 +4350,148 @@ func TestChangefeedEnrichedSourceWithData(t *testing.T) {
+									actualTSNS := actualSource["ts_ns"].(float64)
+									tsnsFloat := strconv.FormatFloat(actualTSNS, 'f', -1, 64)
+									assertReasonableMVCCTimestamp(t, tsnsFloat)
+									actualSource["ts_ns"] = dummyUpdatedTSNS
```
If we're setting both of the actual source and the assertion map in this function all so that we can eventually compare them and make sure they're equal, could it be a little easier and more readable to just assert whatever we want to about them (e.g. with assertReasonableMVCCTimestamp) and then just delete them or not populate them in the assertionMap to begin with?

KeithCh replied:
I'd say deleting would be similar in terms of readability, and you always need to populate the assertion map when using avro.

aerfrei commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -4350,114 +4350,148 @@ func TestChangefeedEnrichedSourceWithData(t *testing.T) {
+								}
+								if withUpdated {
+									tsnsMap := actualSource["source"].(map[string]any)["ts_ns"].(map[string]any)
+									actualTSNS := tsnsMap["long"].(float64)
```
What happens if you cast this to an int64 instead of a float here and format it accordingly? I think this could solve the type issue with assertEqualNSHLCWalltime

KeithCh replied:
You'd get 
> `panic: interface conversion: interface {} is float64, not int64 [recovered]`

asg0451 replied:
the issue here is that json numbers are parsed as floats in go. this can result in loss of precision for int64 numbers. you need to change the bits that does the json decode to use a json Decoder with the UseNumber flag set. see https://pkg.go.dev/encoding/json#example-Decoder and https://pkg.go.dev/encoding/json#Decoder.UseNumber . This way, all numbers will be decoded as `json.Number` which is a newtype wrapper for string. you'll then need to parse that manually (if you want, or leave it in a string if more convenient)

asg0451 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -11019,3 +11053,10 @@ func assertReasonableMVCCTimestamp(t *testing.T, ts string) {
 	require.GreaterOrEqual(t, epochNanos, now.Add(-1*time.Hour).UnixNano())
 }
+
+// tsns is float64 for some reason when reading from enriched metadata
+// source body, keeping the same type here to not lose precision
```
see previous comment

asg0451 reviewed (commented):
looks good except for the test float thing. lmk if you have any qs about that

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -131,18 +139,24 @@ func (p *enrichedSourceProvider) KafkaConnectJSONSchema() kcjsonschema.Schema {
+func (p *enrichedSourceProvider) GetJSON(
+	updated cdcevent.Row, evCtx eventContext,
+) (json.JSON, error) {
 	// TODO(#139661): Add other non fixed fields.
```
can you rm this comment and other ones like it? now that the work is done

asg0451 commented on pkg/ccl/changefeedccl/helpers_test.go:
```diff
@@ -143,8 +143,10 @@ func applySourceAssertion(
+		decoder := gojson.NewDecoder(bytes.NewReader(m.Value))
+		decoder.UseNumber()
+		if err := decoder.Decode(&message); err != nil {
+			return errors.Wrapf(err, `decode: %s`, m.Value)
```
you need to also check that decoder has read all the bytes in the input. one way to do this is to compare https://pkg.go.dev/encoding/json#Decoder.InputOffset with len(m.Value)

asg0451 reviewed (approved):
lgtm with a few nits remaining.

KeithCh commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -4350,114 +4350,150 @@ func TestChangefeedEnrichedSourceWithData(t *testing.T) {
+								if withUpdated {
+									actualTSNS, err := actualSource["ts_ns"].(gojson.Number).Int64()
+									require.NoError(t, err)
+									tsnsString := strconv.FormatInt(actualTSNS, 10)
```
Still have it to cast it to a string but I used the builtin in method `.string()` instead

KeithCh commented:
bors r+
